server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1245,22 +1237,23 @@ private static void validateAliasOrIndex(String expression) {
             }
         }
 
-        private static boolean aliasOrIndexExists(Context context, IndicesOptions options, Metadata metadata, String expression) {
-            IndexAbstraction indexAbstraction = metadata.getIndicesLookup().get(expression);
+        private static Optional<RuntimeException> aliasOrIndexExists(Context context, String expression) {
```

Comment:
I think it's confusing to generate an exception and decide whether or not to throw it later in the code. On the other hand, I see the benefit in reducing the the checks and lookups that were previously duplicated.

I can't think of an obviously better way of doing this. The best I could come up with was:
```
        private static boolean aliasOrIndexExists(Context context, String expression, boolean throwExceptionIfAbsent) {
            final IndicesOptions options = context.getOptions();
            IndexAbstraction indexAbstraction = context.getState().getMetadata().getIndicesLookup().get(expression);
            if (indexAbstraction == null) {
                if (throwExceptionIfAbsent) {
                    throw indexNotFoundException(expression);
                }
                return false;
            }
...
```

What do you think?

Reply:
Agree this is odd pattern. It is also quite expensive to raise a exceptions especially if not used. I think some usage patterns would see a noticeable performance hit. I like William's suggestion.

Reply:
Thanks for th suggestion, I have pushed: https://github.com/elastic/elasticsearch/pull/89000/commits/21ca35cff0a0a5cf0462765c90e970601c581775 .